packages/arb-avm-cpp/data_storage/src/combinedmachinecache.cpp

```diff
@@ -16,21 +16,42 @@
 
 #include <data_storage/combinedmachinecache.hpp>
 
+void CombinedMachineCache::last_add(std::unique_ptr<Machine> machine) {
+    std::unique_lock lock(mutex);
+
+    last_machine = std::move(machine);
+}
+
 void CombinedMachineCache::basic_add(std::unique_ptr<Machine> machine) {
     std::unique_lock lock(mutex);
 
+    if (last_machine != nullptr &&
+        last_machine->machine_state.arb_gas_remaining <
+            machine->machine_state.arb_gas_remaining) {
+        last_machine = nullptr;
+    }
     basic.add(std::move(machine));
 }
 
 void CombinedMachineCache::lru_add(std::unique_ptr<Machine> machine) {
     std::unique_lock lock(mutex);
 
+    if (last_machine != nullptr &&
+        last_machine->machine_state.arb_gas_remaining <
+            machine->machine_state.arb_gas_remaining) {
+        last_machine = nullptr;
+    }
```

Comment:
Probably good to add a helper method for this

Reply:
:+1: